tableauserverclient/server/endpoint/datasources_endpoint.py

```diff
@@ -282,10 +283,34 @@ def publish(
             new_datasource = DatasourceItem.from_response(server_response.content, self.parent_srv.namespace)[0]
             logger.info("Published {0} (ID: {1})".format(filename, new_datasource.id))
             return new_datasource
-        server_response = self.post_request(url, xml_request, content_type)
-        new_datasource = DatasourceItem.from_response(server_response.content, self.parent_srv.namespace)[0]
-        logger.info("Published {0} (ID: {1})".format(filename, new_datasource.id))
-        return new_datasource
+
+    @api(version="3.13")
+    def update_data(self, datasource_or_connection_item, *, request_id, actions, payload = None):
```

Comment:
Can you add some documentation about the parameters and what this method does + links to the documentation of the API?

Reply:
see https://github.com/tableau/server-client-python/pull/893

Reply:
How does this work? Do we split the documentation from the code or is this somehow merged?

Reply:
for TSC, documentation and code are separated.
Not sure why, but that's how it is...
(Personally, I would prefer if functions would be documented in the source code itself, and the docs were generated from the source code, but that's not how TSC is currently handling its documentation)

Reply:
Should this use BackgroundJobItem instead of JobItem?

Reply:
not sure what the difference between `JobItem` and `BackgroundJobItem` is.
@t8y8, @bcantoni which one should I use here?

Reply:
I've been trying to work out if the update_data method should wait for complete or return the jobitem and it looks like returning the jobitem is more consistent as this matches the other async method here (refresh).  

To make this easier to use would you be open to adding a convenience method to the Jobs endpoint to wait for job completion - (e.g. 
`
    @api(version="3.13")
    def wait_for_async_job(self, async_job_id, timeout=None):
        
        ASYNC_JOB_POLL_INTERVAL=30
        completed_at = None
        finish_code = None
        jobinfo = None

        wait_start_time = time.time()
        
        while completed_at is None:
            time.sleep(ASYNC_JOB_POLL_INTERVAL)
            jobinfo = self.get_by_id(async_job_id)
            completed_at = jobinfo.completed_at
            if completed_at is None:
                logger.debug("Job {} ... progress={}".format(async_job_id, jobinfo.progress))
                if timeout is not None:
                    wait_elapsed_time = (time.time() - wait_start_time) * 1000
                    if wait_elapsed_time > timeout:
                        raise TimeoutError(f"Timeout after {wait_elapsed_time} seconds waiting for asynchronous job: {async_job_id}")
            else:
                finish_code = jobinfo.finish_code
            
        logger.info("Job {} Completed: Finish Code: {} - Notes:{}".format(async_job_id, finish_code, jobinfo.notes))
        return finish_code
`

Reply:
Do you think that the need to write the json for the "actions" parameter makes this harder to use than it needs to be? - I think we should keep this method so that you have the flexibility of writing your own 'actions' json but also create some convenience methods to pass the options as parameters - e.g.

    def update_datasource(
        self, datasource_or_connection_item, *, 
        payload, 
        matching_columns=[['ROWID','ROWID']], 
        source_schema="Extract", 
        source_table="Extract", 
        target_schema="Extract", 
        target_table="Extract"
    ):
        match_conditions_args = []
        for match_pair in matching_columns:
            match_conditions_args.append(
                {
                    "op": "eq",
                    "source-col": match_pair[0],
                    "target-col": match_pair[1],
                }
            )
        if len(match_conditions_args) > 1:
            match_conditions_json = {"op": "and", "args": match_conditions_args}
        else:
            match_conditions_json = match_conditions_args[0]

        json_request = [
            # UPDATE action
            {
                "action": "update",
                "source-schema": source_schema,
                "source-table": source_table,
                "target-schema": target_schema,
                "target-table": target_table,
                "condition": match_conditions_json,
            },
        ]

        request_id = str(uuid.uuid4())
        return self.update_data(datasource_or_connection_item, request_id, json_request, payload)

i.e. hide some of the internals and allow an update by just calling:
`server.datasources.update_datasource(datasource_id, path_to_database,[[keycol1,keycol1],[keycol2,keycol2]])`

Reply:
> adding a convenience method to the Jobs endpoint to wait for job completion

Implemented a proposal in #903. Please provide feedback on the proposed `wait_for_job` in that other PR.

> create some convenience methods to pass the options as parameters

I would prefer not to do so. We have a documented and powerful "actions" language, i.e. the JSON structures. We shouldn't invent yet another "wrapper mini-language" here in this Python library.

Common cases (such as a short-hand notation for `matching_columns` which doesn't need explicit `op: eq` elements) should rather be added on the server-side. That way, all clients (also JavaScript clients, etc.) can benefit from it.

(actually, there is a proposal for a `matching_colums` short-hand syntax. It's just that we never implemented it... @jonas-eckhardt do you think we should re-prioritize that convenience feature on the server-side?)